MovingBlocks/Terasology
# OpenVR motion controller gameplay through controller button API and e…

### Contains

Support for gameplay using OpenVR-compatible motion-tracked controllers. The controller can be used to target in-game objects and trigger actions (i.e. mine blocks that you are touching with the controller).

A video is worth a thousand words: https://youtu.be/zkZ04NFy73U

### How to test

If you have a VR headset, start the program, load/create a game, and put the headset on. Put on a controller. Try to do the same thing in the video.

If you do not, try to turn play the game as you normally would. If you notice any problems with the targeting system, it could be a side effect of this commit. (However, I have tested non-VR gameplay, and everything seems to work as normal). Also pay attention to the animations / position of the tool.

A couple of things that I should mention.
* Before this commit, the target of actions and the target as seen by the renderer were being determined by two different instances of ray casting in the gaze direction. I changed it so that the target assigned to both is given by the targeting system instead of being done separately. This made it easier to hijack the targeting system for my own purposes.
  * When VR is being used, the target system is used to assign a target based on the OpenVR controller location.
  * When VR is not being used, the target system follows the gaze location, and the behavior is the same as it used to be (although what is happening behind the scenes is a little more efficient).
* I found some time ago that, when VR mode was active, I was only able to see the world rendered in the headset. The rendering to the computer monitor was not working, because the `CopyImageToHMD` node was binding a different FBO without telling the DAG about it. I made a temporary fix - to bind FBO 0 after `CopyImageToHMD` terminates. It's at least not much uglier than before.
* Controller support is done through the existing controller API. Each controller has a trigger, a grip button, a menu button, and a touchpad with joystick axes. On the left controller, the first three buttons are bound to buttons 0 - 2, and the touchpad is bound to movement. On the second controller, the first three buttons are bound to 3-5 and the four directions of the touchpad are bound to buttons 6-9. I haven't given much thought to what the best layout for gameplay is, but the back-end is in place.
* There are a couple of random VR fixes in there too. Before, the eye height was not being accounted for and the floor plane didn't match the floor plane in the room. That is fixed. There was also a minor problem with the camera that regarding the inverse projection matrix that could have been causing some problems that is now fixed.
